src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs

```diff
@@ -0,0 +1,824 @@
+                    // Listen for a client blob.
+                    StartReceiveBlob(lazyResult);
+                }
+                else
+                {
+                    // Start with the first blob.
+                    StartSendBlob(null, lazyResult);
+                }
+            }
+            catch (Exception e)
+            {
+                // Round-trip it through SetException().
+                e = SetException(e);
+                throw;
+            }
+            finally
+            {
+                if (lazyResult == null || _exception != null)
+                {
+                    _nestedAuth = 0;
+                }
+            }
+        }
+
+        internal void EndProcessAuthentication(IAsyncResult result)
+        {
+            if (result == null)
+            {
+                throw new ArgumentNullException("asyncResult");
+            }
+
+            LazyAsyncResult lazyResult = result as LazyAsyncResult;
+            if (lazyResult == null)
+            {
+                throw new ArgumentException(SR.Format(SR.net_io_async_result, result.GetType().FullName), "asyncResult");
+            }
+
+            if (Interlocked.Exchange(ref _nestedAuth, 0) == 0)
+            {
+                throw new InvalidOperationException(SR.Format(SR.net_io_invalidendcall, "EndAuthenticate"));
+            }
+
+            // No "artificial" timeouts implemented so far, InnerStream controls that.
+            lazyResult.InternalWaitForCompletion();
+
+            Exception e = lazyResult.Result as Exception;
+
+            if (e != null)
+            {
+                // Round-trip it through the SetException().
+                e = SetException(e);
+                throw e;
+            }
+        }
+
+        private bool CheckSpn()
+        {
+            if (_context.IsKerberos)
+            {
+                return true;
+            }
+
+            if (_extendedProtectionPolicy.PolicyEnforcement == PolicyEnforcement.Never ||
+                    _extendedProtectionPolicy.CustomServiceNames == null)
+            {
+                return true;
+            }
+
+            string clientSpn = _context.ClientSpecifiedSpn;
+
+            if (String.IsNullOrEmpty(clientSpn))
+            {
+                if (_extendedProtectionPolicy.PolicyEnforcement == PolicyEnforcement.WhenSupported)
+                {
+                    return true;
+                }
+            }
+            else
+            {
+                return _extendedProtectionPolicy.CustomServiceNames.Contains(clientSpn);
+            }
+
+            return false;
+        }
+
+        //
+        // Client side starts here, but server also loops through this method.
+        //
+        private void StartSendBlob(byte[] message, LazyAsyncResult lazyResult)
+        {
+            Exception exception = null;
+            if (message != s_emptyMessage)
+            {
+                message = GetOutgoingBlob(message, ref exception);
```

Comment:
@shrutigarg Does it mean that in Unix we will use the same NTAuthentication.cs as well to GetOutgoingBlob ?

This NTAuthentication.GetOutgoinBlob essentially what System.Data need to get SSPI message for Integrated feature. Is there any plan to make this available to public ?

Thanks